File: scripts/urlencode.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

char* urlencode(const char* str) {
    char* encoded = malloc(strlen(str) * 3 + 1); // Allocate memory for the encoded string
    if (encoded == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        exit(1);
    }
    int index = 0;
    for (int i = 0; str[i] != '\0'; i++) {
        if (isalnum(str[i]) || strchr(".~-_", str[i]) != NULL) {
            encoded[index++] = str[i];
        } else {
            sprintf(encoded + index, "%%%02X", (unsigned char)str[i]);
            index += 3;
        }
    }
    encoded[index] = '\0'; // Null-terminate the string
    return encoded;
}

int main() {
    char input_string[1000];
    printf("Enter the string you want to URL encode: ");
    fgets(input_string, sizeof(input_string), stdin);
    input_string[strcspn(input_string, "\n")] = '\0'; // Remove trailing newline character
    
    char* encoded_string = urlencode(input_string);
    printf("%s\n", encoded_string);
    free(encoded_string); // Free the dynamically allocated memory
    return 0;
}

File: scripts/urlencode.go
package main

import (
    "fmt"
    "net/url"
    "os"
    "bufio"
)

func main() {
    reader := bufio.NewReader(os.Stdin)
    fmt.Print("Enter the string you want to URL encode: ")
    inputString, _ := reader.ReadString('\n')
    
    encodedString := urlencode(inputString)
    fmt.Println(encodedString)
}

func urlencode(str string) string {
    encoded := ""
    for _, char := range str {
        switch {
        case (char >= 'a' && char <= 'z') || (char >= 'A' && char <= 'Z') || (char >= '0' && char <= '9') || char == '.' || char == '~' || char == '-' || char == '_':
            encoded += string(char)
        default:
            encoded += fmt.Sprintf("%%%02X", char)
        }
    }
    return encoded
}
